solidity/contracts/hooks/CCIPHookPublisher.sol

```diff
@@ -0,0 +1,217 @@
+
+    /// @notice Constructor initializes the contract with the router address.
+    /// @param _router The address of the router contract.
+    constructor(
+        address _router,
+        address _mailbox,
+        uint32 _destinationDomain,
+        bytes32 _ism
+    )
+        AbstractMessageIdAuthHook(_mailbox, _destinationDomain, _ism) {
+        i_router = _router;
+    }
+
+    // ============ Modifiers ============
+
+    /// @dev Modifier that checks if the chain with the given destinationChainSelector is allowlisted.
+    /// @param _destinationChainSelector The selector of the destination chain.
+    modifier onlyAllowlistedDestinationChain(uint64 _destinationChainSelector) {
+        if (!allowlistedDestinationChains[_destinationChainSelector])
+            revert DestinationChainNotAllowlisted(_destinationChainSelector);
+        _;
+    }
+
+    // ============ Internal functions ============
+
+    /// @notice Construct a CCIP message.
+    /// @dev This function will create an EVM2AnyMessage struct with all the necessary information for sending a text.
+    /// @param _receiver The address of the receiver.
+    /// @param _callData The string data to be sent.
+    /// @param _feeTokenAddress The address of the token used for fees. Set address(0) for native gas.
+    /// @return Client.EVM2AnyMessage Returns an EVM2AnyMessage struct which contains information for sending a CCIP message.
+    function _buildCCIPMessage(
+        address _receiver,
+        bytes memory _callData,
+        address _feeTokenAddress
+    ) internal pure returns (Client.EVM2AnyMessage memory) {
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        return
+            Client.EVM2AnyMessage({
+                receiver: abi.encode(_receiver), // ABI-encoded receiver address
+                data: abi.encode(_callData), // ABI-encoded payload
+                tokenAmounts: new Client.EVMTokenAmount[](0), // Empty array aas no tokens are transferred
+                extraArgs: Client._argsToBytes(
+                    // Additional arguments, setting gas limit and non-strict sequencing mode
+                    Client.EVMExtraArgsV1({gasLimit: 200_000, strict: false})
+                ),
+                // Set the feeToken to a feeTokenAddress, indicating specific asset will be used for fees
+                feeToken: _feeTokenAddress
+            });
+    }
+
+    /// @notice Sends data to receiver on the destination chain.
+    /// @notice Pay for fees in native gas.
+    /// @dev Assumes your contract has sufficient native gas tokens.
+    /// @param _destinationChainSelector The identifier (aka selector) for the destination blockchain.
+    /// @param _receiver The address of the recipient on the destination blockchain.
+    /// @param _callData The text to be sent.
+    /// @return messageId The ID of the CCIP message that was sent.
+    function sendMessagePayNative(
```

Comment:
There is something that has to be clarified:
1. `_destinationChainSelector` is `uint64`, while Hyperlane's `_destinationDomain` is `uint32`. One possible solution would be to introduce the mapping in this case.
2. `_recipient` in the Hyperlane Message is the actual receipient of the sent message, while `_receiver` is simply the address of the contract, which will process the message on the destination chain. We could add extra calls to specify destination ccip receiver address per Hook, for example, but we still have to pass this data somehow.

P.S. just a small question
If I want to send a message, not just a payload, shall I then not inherit `AbstractPostDispatchHook` instead of `AbstractMessageIdAuthHook` and implement `_postDispatch`?

Reply:
I also had a thought, it probably doesn't make sense to limit the `_destinationDomain` to single network, as CCIP can obviously communicate with multiple networks, therefore it makes sense to inherit from `AbstractPostDispatchHook`.
If necessary, a mapping of supported Hyperlane destination domains can be added in this scenario

Reply:
Yes AbstractMessageIdAuthHook was meant for single destination chain but since we handle different destination for the same origin with the RoutingHook we don't need to support it here. At least for now.

I also realized sending the message makes it slightly harder to compose different hooks together so we can leave it as sending just the messageId.

Reply:
Updated